src/Agent.Plugins/PipelineCache/PipelineCacheTaskPluginBase.cs

```diff
@@ -26,36 +22,64 @@ public async Task RunAsync(AgentTaskPluginExecutionContext context, Cancellation
         {
             ArgUtil.NotNull(context, nameof(context));
 
+            string saltVariableName = "AZDEVOPS_PIPELINECACHE_SALT";
+            VariableValue salt = context.Variables.GetValueOrDefault(saltVariableName);
+
+            Func<string,string[]> splitIntoSegments = (s) => {
+                var segments = s.Trim().Split(new [] {'|'},StringSplitOptions.RemoveEmptyEntries).Select(segment => segment.Trim());
+                if(salt != null)
+                {
+                    segments = (new [] { $"{saltVariableName}={salt.Value}"}).Concat(segments);
+                }
+                return segments.ToArray();
+            };
+
             string key = context.GetInput(PipelineCacheTaskPluginConstants.Key, required: true);
 
+            string restoreKeysBlock = context.GetInput(PipelineCacheTaskPluginConstants.RestoreKeys, required: false);
+
+            IEnumerable<string> restoreKeysPerKey;
+            if(string.IsNullOrWhiteSpace(restoreKeysBlock))
+            {
+                restoreKeysPerKey = Enumerable.Empty<string>();
+            }
+            else
+            {
+                restoreKeysBlock = restoreKeysBlock.Replace("\r\n", "\n"); //normalize newlines
+                restoreKeysPerKey = restoreKeysBlock.Split(new [] {'\n'}, StringSplitOptions.RemoveEmptyEntries); // split by marker
+                restoreKeysPerKey = restoreKeysPerKey.Select(k => $"{k} | **"); // all restore-only keys are assumed to be wildcards
```

Comment:
Just trying to get my head around the purpose of wildcard segments and their impact on the fingerprint.

Basic questions:

1. Can a restore key have multiple wildcard segments? If so, what does this mean?
2. Does the server have logic that deals with wildcard segments?
3. In what scenarios are wildcards needed? Or to ask another way ... if we didn't support wildcards, what breaks?

Not sure what the new REST API call looks like, but in general, the more options the agent has for controlling cache lookup, the better (the server should be relatively dumb and just honor the request from the client). For example, instead of the server always matching the first fingerprint as "exact" and other fingerprints as "startswith", we let the client control the matching. This gives us the most flexibility on the agent and makes sense from a general REST API perspective.

Conceptually, this is what I would expect the body of the "get cache" `POST` request to look like ...

```json
[  
   {  
      "key":"{a|b|c}"
   },
   {  
      "key":"{a|b}",
      "inexact": true
   },
   {  
      "key":"{a}",
      "inexact": true
   }
]
```

(where `{x|y}` means the fingerprint for the key `x | y`)

Each key is checked in order.

(note: instead of `inexact` ... a `match` property that supports different values might make more sense: `exact`, `startswith`, `somethingelselater`, etc)

... but maybe I am missing something fundamental about wildcards.

Reply:
> Can a restore key have multiple wildcard segments? If so, what does this mean?

Not right now.  The only place '*' is allowed is if the key ends with a "**" segment.  We could add globbing in the future though.  I suppose the logic of "take the newest of anything that matches" would continue, but not sure.

> Does the server have logic that deals with wildcard segments?

It does a recursive search iff the last key segment is **.  If it finds multiple entries it takes the most recently created.

> In what scenarios are wildcards needed? Or to ask another way ... if we didn't support wildcards, what breaks?

We leverage wildcards for inexact matches.

> For example, instead of the server always matching the first fingerprint as "exact" and other fingerprints as "startswith", we let the client control the matching. This gives us the most flexibility on the agent and makes sense from a general REST API perspective.

This is exactly what the server does (great minds 😁 ) but the "startswith xyz" is encoded as a pseudo-glob "xyz|**".

The POST request is logically the same as what you shared, but it is formatted as such (from memory):
```
{
 keys: [
  "a|b|c",
  "a|b|**",
  "a|**"
  ]
}
```
> Each key is checked in order.

Yup!

>(note: instead of inexact ... a match property that supports different values might make more sense: exact, startswith, somethingelselater, etc)

Yes - think of the "wildcard" as "an extreme subset of globbing with a different separator"